File: web/js/nameplate.js
// A little sign staked in a settler's front yard, carrying the session's own name.
// The post and board are flat-shaded wood; the lettering is a canvas texture, because
// text is the one thing the merged vertex-colour geometry cannot draw.
import * as THREE from 'three';

// The frame and legs are shared across every sign — only the lettered face is per-house.
const woodMat = new THREE.MeshStandardMaterial({ color: 0x6b4a2f, flatShading: true, roughness: 0.85 });
const postMat = new THREE.MeshStandardMaterial({ color: 0x5a3c28, flatShading: true, roughness: 0.85 });

const BOARD_W = 0.86;
const BOARD_H = 0.34;
const CANVAS_W = 512;

function draw(text, boardW, boardH, canvasW, band) {
  const CANVAS_H = Math.round(canvasW * boardH / boardW);     // keep the plaque's aspect ratio
  const c = document.createElement('canvas');
  c.width = canvasW;
  c.height = CANVAS_H;
  const g = c.getContext('2d');
  const CANVAS_W = canvasW;

  // a painted plaque: cream field, routed dark border
  g.fillStyle = '#f3e7cf';
  g.fillRect(0, 0, CANVAS_W, CANVAS_H);
  g.fillStyle = '#e8d8b6';
  g.fillRect(0, 0, CANVAS_W, 10);
  g.strokeStyle = '#5a3c28';
  g.lineWidth = 14;
  g.strokeRect(7, 7, CANVAS_W - 14, CANVAS_H - 14);

  // A hamlet's sign carries its own colour as a band across the top, so two neighbours
  // are told apart at a glance. The hue never goes in the geometry - the board itself is
  // one shared mesh.
  if (band != null) {
    g.fillStyle = `hsl(${band} 52% 46%)`;
    g.fillRect(7, 7, CANVAS_W - 14, Math.round(CANVAS_H * 0.12));
  }

  const words = String(text || 'Untitled session').trim().split(/\s+/);
  const maxW = CANVAS_W - 52;
  g.fillStyle = '#3a2a1a';
  g.textAlign = 'center';
  g.textBaseline = 'middle';

  // Fit in one or two lines, shrinking the type until it fits the board.
  const top = band != null ? Math.round(CANVAS_H * 0.12) : 0;
  for (let size = Math.round(62 * (CANVAS_W / 512)); size >= 14; size -= 2) {
    g.font = `600 ${size}px "Iowan Old Style", "Palatino Linotype", Georgia, serif`;
    const lines = wrap(g, words, maxW);
    if (lines.length <= 2 && lines.every((l) => g.measureText(l).width <= maxW)) {
      const lh = size * 1.12;
      const y0 = (CANVAS_H + top) / 2 - (lines.length - 1) * lh / 2;
      lines.forEach((l, i) => g.fillText(l, CANVAS_W / 2, y0 + i * lh));
      break;
    }
  }

  const tex = new THREE.CanvasTexture(c);
  tex.colorSpace = THREE.SRGBColorSpace;
  tex.anisotropy = 4;
  tex.minFilter = THREE.LinearMipmapLinearFilter;
  return tex;
}

// Greedy wrap into at most two lines; a single word longer than the board is truncated.
function wrap(g, words, maxW) {
  const lines = [];
  let line = '';
  for (const w of words) {
    const trial = line ? `${line} ${w}` : w;
    if (g.measureText(trial).width > maxW && line) { lines.push(line); line = w; }
    else line = trial;
    if (lines.length === 1 && g.measureText(line).width > maxW) break;
  }
  if (line) lines.push(line);
  if (lines.length > 2) {
    let last = lines[1];
    while (g.measureText(`${last}…`).width > maxW && last.length > 1) last = last.slice(0, -1);
    return [lines[0], `${last}…`];
  }
  return lines;
}

// A staked yard sign carrying `text`, its face toward local +z (the door/street side).
// `arch` turns the sign into a gateway: the posts stand a road's width apart and carry a
// beam, and the board hangs beneath it, so you read the name as you walk under it rather
// than passing a placard in a field. Signs are not blockers, so walking through needs no
// collision work - the arch only has to be tall enough to look walkable.
export function createNameplate(text, {
  small = false, width = BOARD_W, height = BOARD_H, canvasW = CANVAS_W, band = null,
  height0 = 0.42, posts = 1, arch = 0,
} = {}) {
  const s = small ? 0.7 : 1;
  const group = new THREE.Group();
  const boardY = height0 * s;

  const grow = Math.sqrt(width / BOARD_W);
  const span = arch ? arch / 2 : width * s * 0.42;
  const legH = arch ? boardY + height * s + 0.12 : boardY + 0.04;
  const legs = [];
  const n = arch ? 2 : posts;
  for (let i = 0; i < n; i++) {
    const r = arch ? 0.045 : 0.022 * grow;
    const leg = new THREE.Mesh(new THREE.CylinderGeometry(r, r * 1.25, legH, 6), postMat);
    leg.position.set(n === 1 ? 0 : (i === 0 ? -1 : 1) * span, legH / 2, 0);
    leg.castShadow = true;
    group.add(leg);
    legs.push(leg);
  }
  if (arch) {
    // The beam overhangs its posts a little, the way a real one is pegged on top.
    const beam = new THREE.Mesh(new THREE.BoxGeometry(arch + 0.22, 0.1, 0.13), postMat);
    beam.position.y = legH + 0.05;
    beam.castShadow = true;
    group.add(beam);
  }


  const bw = width * s, bh = height * s;
  const board = new THREE.Mesh(new THREE.BoxGeometry(bw, bh, 0.03), woodMat);
  board.position.y = boardY + bh / 2 - 0.02;
  board.castShadow = true;
  group.add(board);

  const tex = draw(text, width, height, canvasW, band);
  const faceMat = new THREE.MeshBasicMaterial({ map: tex, transparent: true });
  const face = new THREE.Mesh(new THREE.PlaneGeometry(bw * 0.94, bh * 0.86), faceMat);
  face.position.set(0, board.position.y, 0.017);
  group.add(face);

  group.scale.setScalar(s < 1 ? 1 : 1);   // reserved: keep API stable

  return {
    group,
    dispose() {
      for (const leg of legs) leg.geometry.dispose();
      board.geometry.dispose();
      face.geometry.dispose();
      faceMat.dispose();
      tex.dispose();
    },
  };
}

File: godot/src/World/Nameplate.cs
using System;
using Godot;
using Promptholm.Visual;

namespace Promptholm.World;

/// <summary>
/// The little sign staked in a settler's front yard, carrying the session's own name.
///
/// This is what makes the island readable as a place where work happens rather than a village of
/// anonymous boxes: every house is one Claude session, and the board in front of it says which.
/// Ported from the V1 browser island (<c>web/js/nameplate.js</c>), which established the look —
/// cream field, dark routed border, and a band of the district's own hue across the top so two
/// neighbouring hamlets are told apart at a glance.
///
/// Text is a <see cref="Label3D"/> rather than a painted canvas texture. The browser version had
/// to draw into a canvas because merged vertex-colour geometry cannot carry glyphs; here the
/// engine has a text node that renders crisply at any distance and costs one draw call, and
/// <c>DistrictDecorator</c>'s archway plaques already use it, so this stays in the same idiom.
/// </summary>
public static class Nameplate
{
	/// <summary>Board width in metres. Two lines of a settler name fit comfortably.</summary>
	private const float BoardW = 1.05f;

	private const float BoardH = 0.42f;
	private const float BoardThick = 0.05f;

	/// <summary>Height of the board's underside above the ground.</summary>
	private const float BoardBase = 0.52f;

	private const float PostR = 0.045f;

	private static readonly Color Cream = new(0.94f, 0.90f, 0.81f);
	private static readonly Color Ink = new(0.22f, 0.17f, 0.11f);

	/// <summary>
	/// Builds the sign, with the lettered face toward local −Z — the same way a building's door
	/// slot faces, so a plate parented to a house's root reads from the street without any
	/// rotation of its own.
	///
	/// `hue` is the district's colour in degrees, or null for a building that belongs to no
	/// district (the civics), which then simply gets no band.
	/// </summary>
	public static Node3D Build(string text, float? hue = null, float scale = 1.0f)
	{
		var root = new Node3D { Name = "Nameplate" };

		float bw = BoardW * scale;
		float bh = BoardH * scale;
		float baseY = BoardBase * scale;
		float boardMidY = baseY + bh * 0.5f;

		// Two legs rather than one: a single stake reads as a for-sale sign, two as a plaque that
		// belongs to the house behind it.
		float legH = boardMidY + bh * 0.45f;
		float legSpan = bw * 0.38f;
		var legMesh = new CylinderMesh
		{
			TopRadius = PostR * scale,
			BottomRadius = PostR * 1.25f * scale,
			Height = legH,
			RadialSegments = 6,
			Rings = 1,
			Material = Palette.Solid(Palette.Wood, 0.85f),
		};
		foreach (float sx in new[] { -legSpan, legSpan })
		{
			root.AddChild(new MeshInstance3D
			{
				Name = sx < 0 ? "LegL" : "LegR",
				Mesh = legMesh,
				Position = new Vector3(sx, legH * 0.5f, 0.0f),
			});
		}

		var board = new BoxMesh
		{
			Size = new Vector3(bw, bh, BoardThick * scale),
			Material = Palette.Solid(Cream, 0.88f),
		};
		root.AddChild(new MeshInstance3D
		{
			Name = "Board",
			Mesh = board,
			Position = new Vector3(0.0f, boardMidY, 0.0f),
		});

		// The routed border: a slightly larger, darker slab a hair behind the cream face. Cheaper
		// than four rails and reads the same at the distance anyone sees it from.
		var frame = new BoxMesh
		{
			Size = new Vector3(bw + 0.06f * scale, bh + 0.06f * scale, BoardThick * 0.6f * scale),
			Material = Palette.Solid(new Color(0.28f, 0.19f, 0.10f), 0.85f),
		};
		root.AddChild(new MeshInstance3D
		{
			Name = "Frame",
			Mesh = frame,
			Position = new Vector3(0.0f, boardMidY, 0.012f * scale),
		});

		if (hue.HasValue)
		{
			var band = new BoxMesh
			{
				Size = new Vector3(bw * 0.92f, bh * 0.16f, BoardThick * 0.5f * scale),
				Material = Palette.Solid(Color.FromHsv(hue.Value / 360.0f, 0.52f, 0.46f), 0.85f),
			};
			root.AddChild(new MeshInstance3D
			{
				Name = "HueBand",
				Mesh = band,
				Position = new Vector3(0.0f, boardMidY + bh * 0.36f, -BoardThick * 0.35f * scale),
			});
		}

		var label = new Label3D
		{
			Name = "Text",
			Text = Wrap(text),
			FontSize = 28,
			PixelSize = 0.0050f * scale,
			Modulate = Ink,
			HorizontalAlignment = HorizontalAlignment.Center,
			VerticalAlignment = VerticalAlignment.Center,
			// Face −Z, the street side, matching the door slot.
			RotationDegrees = new Vector3(0.0f, 180.0f, 0.0f),
			Position = new Vector3(0.0f, boardMidY - bh * 0.06f, -BoardThick * 0.55f * scale),
		};
		root.AddChild(label);

		return root;
	}

	/// <summary>
	/// Two lines at most, broken at the last space that fits. Settler names are two words
	/// ("Eager Brook"), so this is nearly always a single break in the middle; a one-word name is
	/// left alone and a very long one is cut rather than allowed to overrun the board.
	/// </summary>
	private static string Wrap(string text)
	{
		string t = (text ?? "").Trim();
		if (t.Length == 0) return "Untitled";
		if (t.Length <= 13) return t;

		int mid = t.LastIndexOf(' ', Math.Min(t.Length - 1, 15));
		if (mid <= 0)
			return t.Length > 15 ? t[..14] + "…" : t;

		string first = t[..mid];
		string second = t[(mid + 1)..];
		if (second.Length > 15) second = second[..14] + "…";
		return $"{first}\n{second}";
	}
}
